### ISSUE TITLE
Connecting to local X server fails in Windows Subsystem for Linux (trying to use Unix socket)

### ISSUE THREAD

**Issue by mizzao** [python-xlib/python-xlib#99]
I'm running Ubuntu on Windows with an X server [as detailed here](https://askubuntu.com/questions/827952/a-better-terminal-experience-for-windows-subsystem-for-linuxwsl). Emacs and other Xlib generally work fine. However, apps using python-xlib break.

With `DISPLAY=:0`, the smallest piece of code that breaks is the following:

```python
import Xlib.display

display = Xlib.display.Display()
display.close()
```

with the following error:

```
Traceback (most recent call last):
  File "/home/mao/projects/python-xlib/Xlib/support/unix_connect.py", line 105, in get_socket
    s.connect(address)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "xlib.py", line 3, in <module>
    display = Xlib.display.Display()
  File "/home/mao/projects/python-xlib/Xlib/display.py", line 89, in __init__
    self.display = _BaseDisplay(display)
  File "/home/mao/projects/python-xlib/Xlib/display.py", line 71, in __init__
    protocol_display.Display.__init__(self, *args, **keys)
  File "/home/mao/projects/python-xlib/Xlib/protocol/display.py", line 94, in __init__
    self.socket = connect.get_socket(name, host, displayno)
  File "/home/mao/projects/python-xlib/Xlib/support/connect.py", line 86, in get_socket
    return mod.get_socket(dname, host, dno)
  File "/home/mao/projects/python-xlib/Xlib/support/unix_connect.py", line 107, in get_socket
    raise error.DisplayConnectionError(dname, str(val))
Xlib.error.DisplayConnectionError: Can't connect to display ":0": [Errno 111] Connection refused
```

I've narrowed it down to [this line](https://github.com/python-xlib/python-xlib/blob/bed20b714b999d0dfd0c83e47044d7c5da304da0/Xlib/support/unix_connect.py#L104), which seems to be trying to use a Unix socket that wouldn't exist on WSL. (The X server is running on Windows).

If I try using `DISPLAY=localhost:0` instead, which by that code seems to force a TCP connection, it seems like a connection happens, but I get a different error. I can't really figure out what is causing this.

```
Traceback (most recent call last):
  File "/home/mao/projects/python-xlib/Xlib/display.py", line 224, in __getattr__
    function = self.display_extension_methods[attr]
KeyError: 'ge_add_event_data'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "xlib.py", line 3, in <module>
    display = Xlib.display.Display()
  File "/home/mao/projects/python-xlib/Xlib/display.py", line 127, in __init__
    mod.init(self, info)
  File "/home/mao/projects/python-xlib/Xlib/ext/xinput.py", line 652, in init
    disp.ge_add_event_data(info.major_opcode, device_event, DeviceEventData)
  File "/home/mao/projects/python-xlib/Xlib/display.py", line 227, in __getattr__
    raise AttributeError(attr)
AttributeError: ge_add_event_data
```

Given that apps using Xlib seem to work fine (with both `DISPLAY=:0` and `DISPLAY=localhost:0`), is there some way we can fix the socket code to be more robust in this situation? I tried browsing the Xlib source but couldn't figure out how the socket connections work there. If you can give me some pointers I'm happy to debug further.

**Comment by mizzao** [comment]
Additional sleuthing revealed that really old versions of Xlib use a [similar method](http://www.retro11.de/ouxr/43bsd/usr/src/new/X/Xlib/XOpenDisplay.c.html) of connection via sockets.

However, current versions [use libxcb](https://github.com/mirror/libX11/blob/master/src/OpenDis.c) instead. Seems like [this part of libxcb](https://github.com/stapelberg/libxcb/blob/master/src/xcb_util.c#L411) is what's being called.

**Comment by benoit-pierre** [comment]
> I'm running Ubuntu on Windows with an X server as detailed here.

Which one? The question as 6 answers...

What version of Python are you using? What version of python-xlib? What does `python -c 'import sys; print(sys.platform)'` report?

**Comment by mizzao** [comment]
Sorry, [this answer](https://askubuntu.com/a/867696) (which I wrote). It's just vanilla Xming.

I'm using Python 3.6.3 installed via Anaconda in the Linux environment. python-xlib is checked out from `master`.

```
$ python -c 'import sys; print(sys.platform)'
linux
```

My experience with WSL is that most pure Linux-based stuff just works; the weird cases are crossing filesystems and also things like this X server not being where most Linux apps expect it to be.

**Comment by benoit-pierre** [comment]
Ok, so we could not use an Unix socket when on Windows, now about the second issue: can you try running `python -c 'from Xlib import display; display.Display("localhost:0")'` with this patch:
```diff
 Xlib/display.py | 1 +
 1 file changed, 1 insertion(+)

diff --git i/Xlib/display.py w/Xlib/display.py
index d5fd3fa..dcf11e8 100644
--- i/Xlib/display.py
+++ w/Xlib/display.py
@@ -111,6 +111,7 @@ class Display(object):
         self.extension_event = rq.DictWrapper({})
 
         exts = self.list_extensions()
+        print('server extensions:', exts)
 
         # Go through all extension modules
         for extname, modname in ext.__extensions__:
```

**Comment by mizzao** [comment]
```
$ python -c 'from Xlib import display; display.Display("localhost:0")'
server extensions: ['SHAPE', 'Extended-Visual-Information', 'XInputExtension', 'XTEST', 'BIG-REQUESTS', 'MIT-SUNDRY-NONSTANDARD', 'DEC-XTRAP', 'SYNC', 'XKEYBOARD', 'XC-MISC', 'RECORD', 'DOUBLE-BUFFER', 'XC-APPGROUP', 'SECURITY', 'TOG-CUP', 'GLX', 'SGI-GLX', 'XFIXES', 'RENDER', 'RANDR', 'X-Resource', 'XEVIE', 'DAMAGE']
```

**Comment by benoit-pierre** [comment]
So no `Generic Event Extension` support, that's weird, according to the changelog for older releases, it should be supported: [starting with Xming 7.4.0.1 (Released 21 Jan 2009)](http://www.straightrunning.com/xmingnotes/release4.php).

**Comment by mizzao** [comment]
Turns out Xming has some weird freemium model where the latested publicly available version is 6.9.x. I switched to [VcXsrv](https://sourceforge.net/projects/vcxsrv/) instead and everything seems to work fine with `DISPLAY=localhost:0`.

```
server extensions: ['XINERAMA', 'Generic Event Extension', 'SHAPE', 'XInputExtension', 'XTEST', 'BIG-REQUESTS', 'SYNC', 'XKEYBOARD', 'XC-MISC', 'SECURITY', 'XFIXES', 'XFree86-Bigfont', 'RENDER', 'RANDR', 'Composite', 'DAMAGE', 'MIT-SCREEN-SAVER', 'DOUBLE-BUFFER', 'RECORD', 'DPMS', 'Present', 'X-Resource', 'GLX', 'SGI-GLX']
```

Any ideas on getting `DISPLAY=:0` to work or is that a lost cause?

**Comment by benoit-pierre** [comment]
I guess we could default to 127.0.0.1 when no protocol is specified and the host is not set on Windows:
```diff
 Xlib/support/unix_connect.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git i/Xlib/support/unix_connect.py w/Xlib/support/unix_connect.py
index 3b16d5d..62123f2 100644
--- i/Xlib/support/unix_connect.py
+++ w/Xlib/support/unix_connect.py
@@ -23,6 +23,7 @@ import re
 import os
 import platform
 import socket
+import sys
 
 # FCNTL is deprecated from Python 2.2, so only import it if we doesn't
 # get the names we need.  Furthermore, FD_CLOEXEC seems to be missing
@@ -73,6 +74,9 @@ def get_display(display):
         # Special case: `unix:0.0` is equivalent to `:0.0`.
         if host == 'unix':
             host = ''
+        # Connect to localhost on Windows.
+        elif not host and sys.platform.startswith('win32'):
+            host = '127.0.0.1'
     dno = int(dno)
     if screen:
         screen = int(screen)
```

**Comment by mizzao** [comment]
I'm not sure that would work here. Inside WSL it looks like Linux:

```
In [1]: import sys

In [2]: sys.platform
Out[2]: 'linux'
```

**Comment by benoit-pierre** [comment]
Right, it's faking a Linux machine... Changing the code to match XCB's behavior is a little more involved, can you confirm this work for you: #100.

**Comment by mizzao** [comment]
Yes, that works. Wonderful!

**Comment by Alilino** [comment]
Hello @mizzao & @benoit-pierre , I am currently facing the same problem stated here,
I wanted to use the pyautogui in python3 but I got an error and found out that the Xlib library is not correctly loading. the error is as follows : 

>  File "/home/me/GuiAutomatization/Xlib/display.py", line 224, in __getattr__
>    function = self.display_extension_methods[attr]
> eyError: 'ge_add_event_data'
> 
> uring handling of the above exception, another exception occurred:
> 
> raceback (most recent call last):
>  File "<string>", line 1, in <module>
>  File "/home/me/GuiAutomatization/Xlib/display.py", line 127, in __init__
>    mod.init(self, info)
>  File "/home/me/GuiAutomatization/Xlib/ext/xinput.py", line 652, in init
>    disp.ge_add_event_data(info.major_opcode, device_event, DeviceEventData)
>  File "/home/me/GuiAutomatization/Xlib/display.py", line 227, in __getattr__
>    raise AttributeError(attr)
> ttributeError: ge_add_event_data
> rror in sys.excepthook:
> raceback (most recent call last):
>  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 63, in apport_excepthook
>    from apport.fileutils import likely_packaged, get_recent_crashes
>  File "/usr/lib/python3/dist-packages/apport/__init__.py", line 5, in <module>
>    from apport.report import Report
>  File "/usr/lib/python3/dist-packages/apport/report.py", line 30, in <module>
>    import apport.fileutils
>  File "/usr/lib/python3/dist-packages/apport/fileutils.py", line 23, in <module>
>    from apport.packaging_impl import impl as packaging
>  File "/usr/lib/python3/dist-packages/apport/packaging_impl.py", line 23, in <module>
>    import apt
>  File "/usr/lib/python3/dist-packages/apt/__init__.py", line 34, in <module>
>    apt_pkg.init_config()
> ystemError: E:Opening configuration file /etc/apt/apt.conf - ifstream::ifstream (13: Permission denied)
> 
> riginal exception was:
> raceback (most recent call last):
>  File "/home/me/GuiAutomatization/Xlib/display.py", line 224, in __getattr__
>    function = self.display_extension_methods[attr]
> eyError: 'ge_add_event_data'
> 
> uring handling of the above exception, another exception occurred:
> 
> raceback (most recent call last):
>  File "<string>", line 1, in <module>
>  File "/home/me/GuiAutomatization/Xlib/display.py", line 127, in __init__
>    mod.init(self, info)
>  File "/home/me/GuiAutomatization/Xlib/ext/xinput.py", line 652, in init
>    disp.ge_add_event_data(info.major_opcode, device_event, DeviceEventData)
>  File "/home/me/GuiAutomatization/Xlib/display.py", line 227, in __getattr__
>    raise AttributeError(attr)
> ttributeError: ge_add_event_data

The same error occurs on a red hat Linux server and an ubuntu subsystem on windows 10!
If you could help me out with this I would be very thankful , because I have been hanging at this for days now.

**Comment by poglesbyg** [comment]
What was the fix for this issue?

**Comment by petli** [comment]
@poglesbyg This bug was closed by PR #100, so that would've been the fix. Could you try with the latest version of the library and if there's still a problem create a new issue for it? That makes it easier to track it.
